Retouch GUI rework

Mark-64 opened the pull request:
This is a rework of the GUI of retouch module.
Main changes
- icons moved to cairo paint library
- rework of the icons (consistent hover effect, optical balance, new icon for blur tool)
- scale cursors consistent with the rest of the DT gui
- cut button sensitive only when there is something to cut
- paste button sensitive and armed only when something has been cut
- thicker yellow bar indicating shapes, it was almost invisible
- current scale indicator: cyan dot instead of thin red box for better contrast and readability. Red box over a green background (when merge from scale is activated) could create problems to people with some color blindness (like me)
- leaner and cleaner code

![Screenshot from 2020-06-16 15-25-08](https://user-images.githubusercontent.com/43290988/84791782-5f5c0000-aff3-11ea-93aa-18e0b6f39730.png)

Some of the choices are subjective of course, so feedback is welcome

johnny-bit commented on src/dtgtk/paint.c:
```diff
@@ -2024,6 +2023,215 @@ void dtgtk_cairo_paint_map_pin(cairo_t *cr, gint x, gint y, gint w, gint h, gint
   FINISH
 }
 
+
+
+
+
+
+
+
+
+
```
bunch of empty lines?

Mark-64 replied:
oops, forgot to remove, thanks

TurboGit commented:
As we are at it I would remove the # characters. That's the only module using this, better to be consistent.

Nilvus commented:
> As we are at it I would remove the **#** characters. That's the only module using this, better to be consistent.

I totally agree. Not needed. Hope this will be in 3.2.

Mark-64 commented:
> Edit: 3.4 would be good
Well it's just a gui "retouch" :-), no deep changes. Should be possible to include in 3.2

TurboGit commented:
> Well it's just a gui "retouch" :-) no deep changes. Should be possible to include in 3.2

Almost not deep change but there is many GUI changes that could introduce issues, we have seen that in the past. You know that I'm not shy of introducing things in dt as fast as possible but at some point the stability of the application must be ensured. I have little time for reviewing new things for 3.2 and we still have hard issues like introspection and some GUI glitches.

Mark-64 commented:
Fair enough, I understand.

Mark-64 commented:
With the last touches
![Screenshot from 2020-06-17 12-48-32](https://user-images.githubusercontent.com/43290988/84889498-3ee68100-b099-11ea-81f2-5d252386d2df.png)

aurelienpierre commented:
Please avoid green in GUI controls, as color-blind people will see it yellow-ish. Cyan is also touchy. See:
![image](https://user-images.githubusercontent.com/2779157/84896376-24fe6b80-b0a4-11ea-8549-6bcf678e0130.png)

The only safe colors are yellow-orange, and pure blue. Otherwise, shades of grey should be favored, especially since we want to reduce visual distractions out of the picture frame.

Mark-64 commented:
green was in the initial design, I didn't change that. But thanks for the explanation, I am a bit color blind myself.
I will try to adapt colors

johnny-bit commented:
I *think* that best "guide" for colour usage for gui is actually filmic rgb + tone equalizer as it's not distracting at all and conveys message well enough. plus it's rather on safe side for colour blindness.

Mark-64 commented:
converted to draft to rework colors

Mark-64 commented:
Reworked colors using shades of grey, as suggested.
The only exception is the shapes indicator (small yellow bar). That must be colored as it should be visible over different shades of grey.
The grey dot changes its shade depending on the background in order to keep a good contrast.
The module should be color-blind-safe and non-distracting now

![Screenshot from 2020-06-17 19-03-36](https://user-images.githubusercontent.com/43290988/84927881-c817ab00-b0cd-11ea-9c64-550905f50965.png)

johnny-bit commented:
I like it! Sorry for a little bike-shedding :wink: i wonder if that yellow bar couldn't be changed to `cairo_set_source_rgb(g->cr, 0.75, 0.50, 0.)` (i like that one, despite it's a warning in tone eq, but it's used to nicelly draw stuff in filmic rgb :) )

I also wonder if there are colors that you could pull from `dt_bauhaus_t` structure (those are updated by style). Check: https://github.com/darktable-org/darktable/blob/e567bdc207ce145eefc6b40cc98fd6b81ded7a47/src/bauhaus/bauhaus.h#L224-L228

:)

aurelienpierre commented on src/dtgtk/paint.c:
```diff
@@ -444,6 +444,7 @@ void dtgtk_cairo_paint_eye(cairo_t *cr, gint x, gint y, gint w, gint h, gint fla
 {
   PREAMBLE(1, 0, 0)
 
+  cairo_set_line_width(cr, 0.1);
```
line width should not be used anymore. It it set in the `PREAMBLE` macro in a consistent way for each icon.

Mark-64 commented:
Removed cairo_set_line_width() whenever possible.
Change color of yellow bar to make @johnny-bit happy :-)
That color is a bit less visible on light grey, though.
I've tried to use  dt_bauhaus_t colors, but it's not really possible, because some needed shades are missing, for example a mid gray 0.5/0.5/0.5.
And additionally it would necessary to use a mix of sliders/graph colors which are not meant to go hand in hand, and a change of bauhaus theme would have unwanted implications of retouch.
So what I have done is to match as much as I could the current bauhaus theme, not perfect but good enough IMO.

![Screenshot from 2020-06-19 10-57-22](https://user-images.githubusercontent.com/43290988/85116079-0cfe2780-b21d-11ea-99ca-41a60c16e997.png)

johnny-bit commented:
Wow, Now it looks very nice to me, but then I have good color vision.

Maybe the way to go here would be to also add dedicated color styling in
css so themes could be matched? But otherwise me very likey!

pt., 19 cze 2020 o 11:09 Marco <notifications@github.com> napisał(a):

> Removed cairo_set_line_width() whenever possible.
> Change color of yellow bar to make @johnny-bit
> <https://github.com/johnny-bit> happy :-)
> That color is a bit less visible on light grey, though.
> I've tried to use dt_bauhaus_t colors, but it's not really possible,
> because some needed shades are missing, for example a mid gray 0.5/0.5/0.5.
> And additionally it would necessary to use a mix of sliders/graph colors
> which are not meant to go hand in hand, and a change of bauhaus theme would
> have unwanted implications of retouch.
> So what I have done is to match as much as I could the current bauhaus
> theme, not perfect but good enough IMO.
>
> [image: Screenshot from 2020-06-19 10-57-22]
> <https://user-images.githubusercontent.com/43290988/85116079-0cfe2780-b21d-11ea-99ca-41a60c16e997.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/5497#issuecomment-646526453>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFIX7H5JWBWJ45J3NX3RXMTM5ANCNFSM4N7WWIKA>
> .
>


-- 
Pozdrawiam,
Hubert Kowalski

Mark-64 commented:
> Maybe the way to go here would be to also add dedicated color styling in
> css so themes could be matched? 

@Nilvus do you agree ?

Nilvus commented:
> > Maybe the way to go here would be to also add dedicated color styling in
> > css so themes could be matched?
> 
> @Nilvus do you agree ?

I have good color vision too. I've just test your work here and good work. Orange color is good for me, even in grey theme (very little more discrete than dark ones but good). And orange is safe color, except for maybe around 0.02% people.

So, no need for me to overload code and CSS file for just that. It's all good on all themes.

ff2000 commented:
> Change color of yellow bar to make @johnny-bit happy :-)
> That color is a bit less visible on light grey, though.
> 
> ![Screenshot from 2020-06-19 10-57-22](https://user-images.githubusercontent.com/43290988/85116079-0cfe2780-b21d-11ea-99ca-41a60c16e997.png)

I am red-green-blind and I have issues with this version. The middle one in the light green box is barely visible (dark and bright boxes are fine).
When I stare at the box from working distance the mark starts to disappear.
Therefor a fake line in the box between the left and the middle mark starts to appear, probably triggered by the arrow above the box? (Or my brain wants to make up a continuous line, who knows...)
I have to get close to the monitor to identify the boxes with
 stripes (and still get dizzy).

I have NO issues with the warning sign in tone equalizer (the sign is big and the background is way darker). I don't know if the color to mark clipping in filmic rgb is the same, but there i also can make out the highlighted areas because it is embedded in a continuous curve - not visible as clearly as the EQU sign but I don't get dizzy when staring at it <;)

The previous version with the bright yellow color worked great for me.

I am just an occasional DT-user, got linked here from an RT discussion, so whichever decision you take I am fine. Just wanted to give feedback.

johnny-bit commented:
Thanks!

While I was bikeshedding for colour consistency i do believe that issue is
caused mostly by how contrast affects bar visibility. Do 3rd and 5th box
work for you? Because those 2 look imo great and should fall within range
of acceptable colors, so maaaaybe on very bright background it would be
good to make the bar "pop" more... Or just go with bright yellow for all ;)

niedz., 21 cze 2020 o 07:02 ff2000 <notifications@github.com> napisał(a):

> Change color of yellow bar to make @johnny-bit
> <https://github.com/johnny-bit> happy :-)
> That color is a bit less visible on light grey, though.
>
> [image: Screenshot from 2020-06-19 10-57-22]
> <https://user-images.githubusercontent.com/43290988/85116079-0cfe2780-b21d-11ea-99ca-41a60c16e997.png>
>
> I am red-green-blind and I have issues with this version. The middle one
> in the light green box is barely visible (dark and bright boxes are fine).
> When I stare at the box from working distance the mark starts to disappear.
> Therefor a fake line in the box between the left and the middle mark
> starts to appear, probably triggered by the arrow above the box? (Or my
> brain wants to make up a continuous line, who knows...)
> I have to get close to the monitor to identify the boxes with
> stripes (and still get dizzy).
>
> I have NO issues with the warning sign in tone equalizer (the sign is big
> and the background is way darker). I don't know if the color to mark
> clipping in filmic rgb is the same, but there i also can make out the
> highlighted areas because it is embedded in a continuous curve - not
> visible as clearly as the EQU sign but I don't get dizzy when staring at it
> <;)
>
> The previous version with the bright yellow color worked great for me.
>
> I am just an occasional DT-user, got linked here from an RT discussion, so
> whichever decision you take I am fine. Just wanted to give feedback.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/5497#issuecomment-647079435>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFL7UAOKYICB5BRBIM3RXWH7FANCNFSM4N7WWIKA>
> .
>


-- 
Pozdrawiam,
Hubert Kowalski

Nilvus commented:
> Or just go with bright yellow for all ;)

Probably the best choice here. It's all about contrast here and yellow will be more contrasted on each grey level. Or an orange more brighter (more "yellowish"!)

ff2000 commented:
> Do 3rd and 5th box work for you?

As I am not sure we count boxes the same way here some notes:
![DT_retouch_gui](https://user-images.githubusercontent.com/1973708/85222962-a8cb9700-b3c7-11ea-9bc4-775b8176a313.png)
The "false bar" occasionally appears when staring at "not ok". The last box looks highlighted differently, don't know if that could confuse users. I do not know the module, just looking at the screenshot, so I have no idea if there is a better way to display the different objects.
How many shades of gray do you need? Are boxes a map from the bar below, so the grays have to match exactly the ones below? If not probably reduce the contrast, so the left box isn't as black and the right box isn't as white, so you can easily find a shapes color to decorate the box that can be seen easily (good brightness contrast) and that your brain doesn't change into a different color.
(for reference concerning color change: 
https://www.moillusions.com/color-tile-illusion-new-aspect/

![lottocube](https://1.bp.blogspot.com/_cxmptAPYR-s/R6Eck042q1I/AAAAAAAAB-g/te9rUJlV2bQ/s1600/lottocube01.jpg)

Mark-64 commented:
@ff2000 I am also red-green-blind, perhaps a bit less than you, as I do not see a false bar and can distinguish the orange bar over the light grey, although not as clearly as the bright yellow.
Yes the orange bar looks different on the white box; but I find a yellow box less visible on white.
We need four shades of grey, but the first and the last have to stand apart as they mean the original and the residual image. I can try to make them a bit less "extreme" and chenge the orange into a more yellowish bright tint.

ff2000 commented:
@Mark-64 how do you like this?
![DT_retouch_gui_mockup](https://user-images.githubusercontent.com/1973708/85254342-bb49dd00-b468-11ea-88f4-4d121e2a63c3.png)
The idea is a uniform white bar with the shape mark sitting inside so there is a better contrast.
This is a quick mockup in inkscape, so big room for improvements ;)

I now also tried it live and found two issues:
1) dragging the scale arrows lags quite a bit
2) your cursor calculations are off ;)
Hovering over the left part of a box highlights the box left to the one I am hovering.
The offset increases with the boxes further to the right.
I assume this has to do with the borders (different width in painting and mouse event functions?) and/or rounding.

ff2000 commented:
>     2\. your cursor calculations are off ;)

The obvious patch:
``` diff
diff --git a/src/iop/retouch.c b/src/iop/retouch.c
index 26ed3e55d..545944f69 100644
--- a/src/iop/retouch.c
+++ b/src/iop/retouch.c
@@ -1061,7 +1061,7 @@ static gboolean rt_wdbar_motion_notify(GtkWidget *widget, GdkEventMotion *event,
   GtkAllocation allocation;
   gtk_widget_get_allocation(widget, &allocation);
   const int inset = round(RT_WDBAR_INSET * allocation.height);
-  const float box_w = allocation.width / (float)RETOUCH_NO_SCALES;
+  const float box_w = (allocation.width - 2.0f * inset) / (float)RETOUCH_NO_SCALES;
 
   /* record mouse position within control */
   g->wdbar_mouse_x = CLAMP(event->x - inset, 0, allocation.width - inset);
```

Mark-64 commented:
@ff2000 
thanks for spotting the mistake in mouse coordinates
I am investigating on the lag you noticed.
See this modification in colors
![image](https://user-images.githubusercontent.com/43290988/85308831-076d3f80-b4b2-11ea-9a5b-54f31a147e10.png)

although your idea of indicator outside the boxes is good because the background would be the same
I have to think about it

Nilvus commented:
@ff2000: for me your mockup is not good. The white line is too visible and add infos not necessary as it's applies to all scales, even if there's no scale activated. We need to not add distraction on UI.

@Mark-64: your last screenshot is the way to go I think. I like that and yellow-orange color seems good for me. And better to previous one as it's more contrasted. I would just lighten a little grey scales (from 3 to last one), just a little (too dark for me). About @ff2000 idea, maybe try to see what we can have if just applied a default top line only on activated scales but something not too bright (so not too white) to avoid distraction. That's part not used so need to be a sort of disabled effect, so something visible but discrete.

ff2000 commented:
> I would just lighten a little grey scales (from 3 to last one), just a little (too dark for me)

Agreed. Now the most bright causes difficulties for me, contrast too low.
I changed `GdkRGBA residual    = {.75, .75, .75, 1};` and it is better for me.
![Screenshot_20200622_192912](https://user-images.githubusercontent.com/1973708/85311832-aef06f00-b4be-11ea-9d25-76d7c51c7453.jpg)
.8 even better, and the other gray tones still look great.

@Nilvus I agree that the bar is too bright. As I said it was just an idea to completely get away from problematic colors and contrasts. Let's see how @Mark-64 decides. The current implementation for sure can still be finetuned.

Mark-64 commented:
@Nilvus 
if we do an outside line only on scales with shapes, it can't be in the top part, or it would overlap with the marker, which is already partially overlapped with the scales visibility grey bar . The alternative would be to raise the marker, but then we would have a marker that normally doesn't touch the boxed, which would be odd.
Better would be to move those indicators at the bottom, below the marker.

@ff2000 I see the same lag in the current master, but yes it lags when dragging.
There are too many updates during the dragging, I think we have to implement a deferred update, but I am bit afraid to touch this. I will try to copy from other parts of DT.

ff2000 commented:
>  I see the same lag in the current master, but yes it lags when dragging.

I didn't try master. I tried last release and that was snappy. And I can't test it because I clicked upgrade DB without thinking :( GRML...
If it's also in master it has nothing to do with this PR and should go in another PR. Focus on polishing the UI and get it merged!

Nilvus commented:
> @Nilvus
> if we do an outside line only on scales with shapes, it can't be in the top part, or it would overlap with the marker, which is already partially overlapped with the scales visibility grey bar . The alternative would be to raise the marker, but then we would have a marker that normally doesn't touch the boxed, which would be odd.
> Better would be to move those indicators at the bottom, below the marker.

As you want. It's good as it is now, except as I said on my previous comment to lighten a little grey scales.

Mark-64 commented:
Just for discussion, I did not commit yet those changes:
- brightened the lighter grey shades
- replaced the orange bar with a grey one, placed at the bottom, visibility is ok now, but perhaps it can be more difficult for the user to remember which bar means what
- avoid overlap of the bars with markers (this can be applied also in the other version
![image](https://user-images.githubusercontent.com/43290988/85369127-c9f9c800-b52c-11ea-8ee7-5955be51d4b4.png)
Now perhaps the boxes are too short, can be made taller of course, although I think it's not needed.

ff2000 commented:
@Mark-64 I like it, looks clean, symmetrical, non-distracting.
To avoid confusion with the top bar you could draw the shapes indicators a little bit shorter than the boxes' widths, so that there always is a little gap between them, avoiding building a continuous line.

(The arrows now aren't centered with the box anymore, but I'm sure you are aware of that :))

johnny-bit commented:
Just a heads up - in 163e260 you commited changes against rawspeed submodule.

Mark-64 commented:
> Just a heads up - in 163e260 you commited changes against rawspeed submodule.

Oops... will try to revert those

Mark-64 commented:
"Final" version
![Screenshot from 2020-06-23 11-44-49](https://user-images.githubusercontent.com/43290988/85390359-04715e00-b549-11ea-811e-168be772c9ba.png)
It seems we are condemned to have color only in our lovely pictures ;-)
I hope this will make everyone happy.
Ready for merge.

LebedevRI commented:
> > Just a heads up - in [163e260](https://github.com/darktable-org/darktable/commit/163e2600bb84956539d99a07c40040327e6d7270) you commited changes against rawspeed submodule.
> 
> Oops... will try to revert those

Please just rebase the changes to not include unwanded back-and-forth tugging.
I never understand why this github's approach is seen as sane..

Mark-64 commented:
> Please just rebase the changes to not include unwanded back-and-forth tugging.

Done, hopefully.

Mark-64 commented:
@LebedevRI 
I squashed the commits doing back and forth changes in rawspeed, rebased and force-pushed.
Now GitHub still see changes in rawspeed, however when you go to see the change it says "nothing to compare, branches up to date"
Any clue ? Sorry for the inconvenience

LebedevRI commented:
(make a backup first)
```
git fetch all -p
git rebase upstream/master master # upstream == https://github.com/darktable-org/darktable
git push -f origin master # to https://github.com/Mark-64/darktable
git rebase master retouch_gui_rework_2
git push -f origin retouch_gui_rework_2
```

Mark-64 commented:
Nope, same problem.
Although I am not sure it is a problem, the diff doesn't show anything

ff2000 commented:
Does ``git submodule update`` help?

johnny-bit commented:
Not necessarilly. You honestly can remove things from commit, so you need to find offending commit and just remove from commit offending parts.

Mark-64 commented:
Victory !!!!
After a long battle, I finally managed to remove rawspeed form the commits :-)

ff2000 commented:
Awesome!
One reminder: Tooltip needs to be adjusted when reviewers decided on the final look.

Nilvus commented:
@Mark-64: Just tested last changes and I'm not really convinced by last changes on scales. Top line and bottom line make some strange "sandwich" effect when having corrections on multiple scales. And make also scales line much higher than before. And it's worse when having a correction on last scale where the bottom arrow is. Color was better for me, as it's more visible and not have same color like top bar.

![2020-06-24_11-33](https://user-images.githubusercontent.com/45535283/85530850-8c6d6b80-b60e-11ea-9db0-c05b6e7b267f.png)

An idea to test : just move bottom indicators just below scales, so between scales and the bottom arrow. They will all be on black background (and try with yellow/orange color as it probably be better view that way for people like @ff2000 and will also remove sandwich effect).

Mark-64 commented:
> just move bottom indicators just below scales

This is the result

![image](https://user-images.githubusercontent.com/43290988/85589167-a2e1ea00-b643-11ea-995e-8a49f59063df.png)

Visibility is good, perhaps the only defect is that the bottom marker does not touch the scales box to leave room for the orange line. I would like those to overlap.
Is that ok ?

Nilvus commented:
> > just move bottom indicators just below scales
> 
> This is the result
> 
> ![image](https://user-images.githubusercontent.com/43290988/85589167-a2e1ea00-b643-11ea-995e-8a49f59063df.png)
> 
> Visibility is good, perhaps the only defect is that the bottom marker does not touch the scales box to leave room for the orange line. I wouldn't like if those overlap.
> Is that ok ?

For me, it's really better than all mockups before. The fact that the indicator doesn't touch the scales box is better I think, for the visibility. And probably better for those who have a not complete color vision, like @ff2000 (waiting for your feedback).
So, it's really good that way.

Did you check how it render for the scale where the bottom arrow is and if space/visibility to scale/indicator/arrow is good between them?

Mark-64 commented:
This is the case you refer to

![image](https://user-images.githubusercontent.com/43290988/85608363-4687c600-b655-11ea-9e9e-d77908621bc5.png)

Nilvus commented:
> This is the case you refer to
> 
> ![image](https://user-images.githubusercontent.com/43290988/85608363-4687c600-b655-11ea-9e9e-d77908621bc5.png)

Yes, so all good for me that way. Thanks

ff2000 commented:
> @Mark-64:Top line and bottom line make some strange "sandwich" effect when having corrections on multiple scales. And make also scales line much higher than before. And it's worse when having a correction on last scale where the bottom arrow is. Color was better for me, as it's more visible and not have same color like top bar.
> 
> ![2020-06-24_11-35](https://user-images.githubusercontent.com/45535283/85531235-dfdfb980-b60e-11ea-9e64-050ede00d44b.png)
The biggest issue in your screenshot is that the bottom and top lines are barely visible as they have a similar grey value as the background of the light theme, and the widget doesn't have a spacing.
Similar issue when the top arrow is below the line. Outlines of the arrows also blend with the background of the light theme.
Would moving the top bar below the top arrows be an issue? (Does the sandwich effect reappear?) And/or add a one pixel spacing to the top and bottom? So that some black from the widget BG is between the indicators and the theme background?

That's also true for the most recent version.
Besides that everything seems to work fine for me, I can see the colored indicators.

Nilvus commented:
> // edit:
> ![85531235-dfdfb980-b60e-11ea-9e64-050ede00d44b](https://user-images.githubusercontent.com/1973708/85613716-d4b67900-b662-11ea-9529-5379df3667a4.png)
> Is it just an arrow? Or an arrow + line. On the bottom it's fixed now, but on the top it's still there.

So with you question, you confirm the problem (it's arrow + line). But all that is no more a problem since last proposal.
